fix(common): expand type for "ngForOf" input to work with strict null checks

devversion opened the pull request:
Currently the `ngForOf` input accepts `null` or `undefined` as valid values. Though
when using strict template input type checking (which will be supported by `ngtsc`),
passing `null` or `undefined` with strict null checks enabled causes a type check
failure because the type for the `ngForOf` input becomes too strict if strict null checks
are enabled. The type of the input needs to be expanded to also accept `null` or
`undefined` to behave consistent (basically regardless of the `strictNullChecks` flag).

This is necessary because whenever strict input type checking is enabled
by default, most of the Angular projects that use `*ngFor` with the async pipe
will either need to disable template type checking or strict null checks because
the `async` pipe returns `null` if the observable hasn't been emitted yet.

See for example how this affects the `angular/components` repository and how
verbose it is to work around it: https://github.com/angular/components/pull/16373.

trotyl commented on packages/core/src/change_detection/differs/iterable_differs.ts:
```diff
@@ -34,7 +34,7 @@ export interface IterableDiffer<V> {
    * `diff()` invocation.
    */
-  diff(object: NgIterable<V>): IterableChanges<V>|null;
+  diff(object: NgIterable<V>|undefined|null): IterableChanges<V>|null;
```
This is breaking change, differs are designed to be extensible and user can make their own differ implementation.

devversion replied:
Yeah we still need evaluate how this affects consumers. This will only be a breaking change for developers that implement an `IterableDiffer` with strict null checks enabled.

This change is fine to land in a major version. Also I think we could work around this and leave the `IterableDiffer` interface untouched.

kara reviewed (approved):
LGTM, but this looks like a breaking change. @IgorMinar will need to weigh in.

devversion committed:
fix(common): expand type for "ngForOf" input to work with strict null checks

Currently the `ngForOf` input accepts `null` or `undefined` as valid
values. Although when using strict template input type checking
(which will be supported by `ngtsc`), passing `null` or `undefined`
with strict null checks enabled causes a type check failure because
the type for the `ngForOf` input becomes too strict if strict null checks
are enabled. The type of the input needs to be expanded to also accept
`null` or `undefined` to behave consistently regardless of the
`strictNullChecks` flag.

This is necessary because whenever strict input type checking is enabled
by default, most of the Angular projects that use `*ngFor` with the async pipe
will either need to disable template type checking or strict null checks
because the `async` pipe returns `null` if the observable hasn't been
emitted yet.

See for example how this affects the `angular/components` repository and
how much bloat the workaround involves: https://github.com/angular/components/pull/16373/files#r296942696.

Splaktar commented:
This was reported in https://github.com/angular/angular/issues/32051.

IgorMinar commented on tools/public_api_guard/core/core.d.ts:
```diff
@@ -506,7 +506,7 @@ export interface IterableChanges<V> {
 
 export interface IterableDiffer<V> {
-    diff(object: NgIterable<V>): IterableChanges<V> | null;
+    diff(object: NgIterable<V> | undefined | null): IterableChanges<V> | null;
```
I'm slightly concerned about this change, but it is a correct change since the differ should accept `null` or `undefined`. Let's presubmit this in g3 and we'll see if this change is ok.

IgorMinar commented:
merge-assistance: please presubmit and check for typing errors due to a slight change of the interface.

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:273204962:BASE:273204996:1570415315328:57d2f33d)
- [VE Global Presubmit](https://test.corp.google.com/OCL:273204962:BASE:273205065:1570415362667:d3853bd)

AndrewKushnir commented:
FYI, VE presubmits (including a [global one](https://test.corp.google.com/ui#id=OCL:273204962:BASE:273304420:1570464282309:6b568857)) are successful. Thank you.

AndrewKushnir commented:
@IgorMinar since this is a breaking change, should we merge it to master only (currently the target is set to master & patch)?

kara commented:
@AndrewKushnir Yes, it should be master-only.

devversion commented:
Yeah, not sure why I marked it as `master & patch.. the PR was pending quite some time 😄
